File: go/pqctls/sig_algorithms.go
package pqctls

import (
	"github.com/baron-chain/PQC-Standards-Implementation/go/composite"
	"github.com/baron-chain/PQC-Standards-Implementation/go/mldsa"
)

// SignatureAlgorithm is a TLS 1.3 signature algorithm identifier for PQC.
type SignatureAlgorithm uint16

const (
	// MLDSA44 is ML-DSA-44 (code point 0x0904).
	MLDSA44 SignatureAlgorithm = 0x0904
	// MLDSA65 is ML-DSA-65 (code point 0x0905).
	MLDSA65 SignatureAlgorithm = 0x0905
	// MLDSA87 is ML-DSA-87 (code point 0x0906).
	MLDSA87 SignatureAlgorithm = 0x0906
	// MLDSA65Ed25519 is ML-DSA-65 + Ed25519 composite (code point 0x0907).
	MLDSA65Ed25519 SignatureAlgorithm = 0x0907
	// MLDSA87Ed25519 is ML-DSA-87 + Ed25519 composite (code point 0x0908).
	MLDSA87Ed25519 SignatureAlgorithm = 0x0908
)

// AllSignatureAlgorithms returns all defined PQC signature algorithms.
var AllSignatureAlgorithms = []SignatureAlgorithm{MLDSA44, MLDSA65, MLDSA87, MLDSA65Ed25519, MLDSA87Ed25519}

// String returns a human-readable name.
func (sa SignatureAlgorithm) String() string {
	switch sa {
	case MLDSA44:
		return "MLDSA44"
	case MLDSA65:
		return "MLDSA65"
	case MLDSA87:
		return "MLDSA87"
	case MLDSA65Ed25519:
		return "MLDSA65_ED25519"
	case MLDSA87Ed25519:
		return "MLDSA87_ED25519"
	default:
		return "Unknown"
	}
}

// IsComposite returns true if the algorithm is a composite (hybrid) signature.
func (sa SignatureAlgorithm) IsComposite() bool {
	return sa == MLDSA65Ed25519 || sa == MLDSA87Ed25519
}

// SignatureAlgorithmFromCodePoint looks up a signature algorithm by code point.
func SignatureAlgorithmFromCodePoint(cp uint16) (SignatureAlgorithm, bool) {
	switch SignatureAlgorithm(cp) {
	case MLDSA44, MLDSA65, MLDSA87, MLDSA65Ed25519, MLDSA87Ed25519:
		return SignatureAlgorithm(cp), true
	default:
		return 0, false
	}
}

func mldsaParams(alg SignatureAlgorithm) *mldsa.Params {
	switch alg {
	case MLDSA44:
		return mldsa.MLDSA44
	case MLDSA65:
		return mldsa.MLDSA65
	case MLDSA87:
		return mldsa.MLDSA87
	default:
		panic("pqctls: not a pure ML-DSA algorithm")
	}
}

func compositeScheme(alg SignatureAlgorithm) composite.Scheme {
	switch alg {
	case MLDSA65Ed25519:
		return composite.MlDsa65Ed25519
	case MLDSA87Ed25519:
		return composite.MlDsa87Ed25519
	default:
		panic("pqctls: not a composite algorithm")
	}
}

// SigningKeyPair holds a signing key pair.
type SigningKeyPair struct {
	PK        []byte
	SK        []byte
	Algorithm SignatureAlgorithm
}

// GenerateSigningKey generates a signing key pair.
func GenerateSigningKey(alg SignatureAlgorithm) *SigningKeyPair {
	switch alg {
	case MLDSA44, MLDSA65, MLDSA87:
		pk, sk := mldsa.KeyGen(mldsaParams(alg))
		return &SigningKeyPair{PK: pk, SK: sk, Algorithm: alg}

	case MLDSA65Ed25519, MLDSA87Ed25519:
		kp := composite.KeyGen(compositeScheme(alg))
		return &SigningKeyPair{PK: kp.PK, SK: kp.SK, Algorithm: alg}

	default:
		panic("pqctls: unsupported signature algorithm")
	}
}

// SignHandshake signs a TLS 1.3 CertificateVerify handshake hash.
func SignHandshake(alg SignatureAlgorithm, sk, handshakeHash []byte) []byte {
	switch alg {
	case MLDSA44, MLDSA65, MLDSA87:
		return mldsa.Sign(sk, handshakeHash, mldsaParams(alg))

	case MLDSA65Ed25519, MLDSA87Ed25519:
		scheme := compositeScheme(alg)
		kp := &composite.KeyPair{PK: nil, SK: sk, Scheme: scheme}
		return composite.Sign(kp, handshakeHash)

	default:
		panic("pqctls: unsupported signature algorithm")
	}
}

// VerifyHandshake verifies a TLS 1.3 CertificateVerify signature.
func VerifyHandshake(alg SignatureAlgorithm, pk, handshakeHash, signature []byte) bool {
	switch alg {
	case MLDSA44, MLDSA65, MLDSA87:
		return mldsa.Verify(pk, handshakeHash, signature, mldsaParams(alg))

	case MLDSA65Ed25519, MLDSA87Ed25519:
		return composite.Verify(compositeScheme(alg), pk, handshakeHash, signature)

	default:
		return false
	}
}

File: python/pqctls/sig_algorithms.py
"""PQC Signature Algorithms for TLS 1.3.

Defines PQC and composite signature algorithms for the signature_algorithms
extension (CertificateVerify), along with sign/verify helpers.
"""

import sys
import os
from dataclasses import dataclass
from enum import IntEnum
from typing import Optional

sys.path.insert(0, os.path.join(os.path.dirname(__file__), ".."))

from mldsa.dsa import keygen as mldsa_keygen, sign as mldsa_sign, verify as mldsa_verify
from mldsa.params import ML_DSA_44, ML_DSA_65, ML_DSA_87
from composite.composite_sig import (
    CompositeScheme, CompositeKeyPair,
    MLDSA65_ED25519, MLDSA87_ED25519,
    key_gen as composite_keygen,
    sign as composite_sign,
    verify as composite_verify,
)


class SignatureAlgorithm(IntEnum):
    """TLS 1.3 signature algorithm identifiers for PQC."""
    MLDSA44 = 0x0904
    MLDSA65 = 0x0905
    MLDSA87 = 0x0906
    MLDSA65_ED25519 = 0x0907
    MLDSA87_ED25519 = 0x0908


ALL_SIGNATURE_ALGORITHMS = list(SignatureAlgorithm)


def signature_algorithm_from_code_point(cp: int) -> Optional[SignatureAlgorithm]:
    """Look up a signature algorithm by its TLS code point."""
    try:
        return SignatureAlgorithm(cp)
    except ValueError:
        return None


def _mldsa_params(alg: SignatureAlgorithm):
    if alg == SignatureAlgorithm.MLDSA44:
        return ML_DSA_44
    elif alg == SignatureAlgorithm.MLDSA65:
        return ML_DSA_65
    elif alg == SignatureAlgorithm.MLDSA87:
        return ML_DSA_87
    raise ValueError(f"Not a pure ML-DSA algorithm: {alg}")


def _composite_scheme(alg: SignatureAlgorithm) -> CompositeScheme:
    if alg == SignatureAlgorithm.MLDSA65_ED25519:
        return MLDSA65_ED25519
    elif alg == SignatureAlgorithm.MLDSA87_ED25519:
        return MLDSA87_ED25519
    raise ValueError(f"Not a composite algorithm: {alg}")


def is_composite(alg: SignatureAlgorithm) -> bool:
    """Whether the algorithm is a composite (hybrid) signature."""
    return alg in (SignatureAlgorithm.MLDSA65_ED25519, SignatureAlgorithm.MLDSA87_ED25519)


@dataclass
class SigningKeyPair:
    """A signing key pair for a PQC signature algorithm."""
    pk: bytes
    sk: bytes
    algorithm: SignatureAlgorithm


def generate_signing_key(alg: SignatureAlgorithm) -> SigningKeyPair:
    """Generate a signing key pair for the given signature algorithm."""
    if is_composite(alg):
        scheme = _composite_scheme(alg)
        kp = composite_keygen(scheme)
        return SigningKeyPair(pk=kp.pk, sk=kp.sk, algorithm=alg)
    params = _mldsa_params(alg)
    pk, sk = mldsa_keygen(params)
    return SigningKeyPair(pk=pk, sk=sk, algorithm=alg)


def sign_handshake(alg: SignatureAlgorithm, sk: bytes, handshake_hash: bytes) -> bytes:
    """Sign a TLS 1.3 CertificateVerify handshake hash."""
    if is_composite(alg):
        scheme = _composite_scheme(alg)
        kp = CompositeKeyPair(pk=b"", sk=sk, scheme=scheme)
        return composite_sign(kp, handshake_hash)
    params = _mldsa_params(alg)
    return mldsa_sign(sk, handshake_hash, params)


def verify_handshake(
    alg: SignatureAlgorithm,
    pk: bytes,
    handshake_hash: bytes,
    signature: bytes,
) -> bool:
    """Verify a TLS 1.3 CertificateVerify signature."""
    if is_composite(alg):
        scheme = _composite_scheme(alg)
        return composite_verify(scheme, pk, handshake_hash, signature)
    params = _mldsa_params(alg)
    return mldsa_verify(pk, handshake_hash, signature, params)
